src/NSubstitute/Proxies/CastleDynamicProxy/CastleDynamicProxyFactory.cs

```diff
@@ -81,23 +129,20 @@ private static bool IsNotCallRouterMethod(MethodInfo methodInfo)
 
             private static bool IsNotBaseObjectMethod(MethodInfo methodInfo)
             {
-                return methodInfo.GetBaseDefinition().DeclaringType != typeof (object);
+                return methodInfo.GetBaseDefinition().DeclaringType != typeof(object);
             }
         }
 
-        private void VerifyNoConstructorArgumentsGivenForInterface(object[] constructorArguments)
+        private class ProxyIdGenerator
         {
-            if (constructorArguments != null && constructorArguments.Length > 0)
-            {
-                throw new SubstituteException("Can not provide constructor arguments when substituting for an interface.");
-            }
-        }
+            private long _idCounter;
 
-        private void VerifyClassHasNotBeenPassedAsAnAdditionalInterface(Type[] additionalInterfaces)
-        {
-            if (additionalInterfaces != null && additionalInterfaces.Any(x => x.GetTypeInfo().IsClass))
+            public string GenerateId(Type primaryProxyType)
             {
-                throw new SubstituteException("Can not substitute for multiple classes. To substitute for multiple types only one type can be a concrete class; other types can only be interfaces.");
+                var id = Interlocked.Increment(ref _idCounter);
+                var shortTypeName = primaryProxyType.GetNonMangledTypeName();
+
+                return string.Format(CultureInfo.InvariantCulture, "Substitute.{0}#{1}", shortTypeName, id);
```

Comment:
In fact, it gives `Substitute.IFoo#1`. Given that `#` is the prohibited character for type in C#, that shouldn't confuse the people.

> I don't think I can come up with anything better (Substitute#{1} [{0}]?)

My goal was to make a substitute id as concise as possible. Some test frameworks have limits on the number of characters they show in e.g. error messages. Therefore the shorter the ID is, the more information will fit into the screen/message. Additionally, it's better to not have spaces in the ID - we never know how they will be formatted).

Personally, I feel that `Substitute.type#id` is the best tradeoff between readability and length.

> I wonder if we could simplify this but still meet the immediate need by switching to Java-style Object.toString which just shoves the hashCode on the end? So for Substitute.For<IFoo>() we'd get Substitute.IFoo@{hashCode} (or hashcode in hex). This may help avoid people getting confused trying to assign meaning to the sequence number generated (as per the IFoo#1/IBar#2 comment).

Just realized that you are completely right. This counter might look strange to people, as it's global across all the tests and they might expect that counter is local to the proxy type. If we stick to the hash code, we'll likely always have large number, so it will be clear to user that numbers are non-predictable. I've changed the code to fix that.

Now about the format. I see a few options:
- `Substitute.IInterface#007877b1` - padded hex
- `Substitute.IInterface#13b9a` - non padded hex
- `Substitute.IInterface#59207139` - non-padded decimal

I suggest us to stick to hex, as hash code might be negative and will look ugly in decimal: `Substitute.IInterface#-7894961`. Also the perfectionist inside me votes for padding 😅 

As for the delimiter, I see the following options:
- `Substitute.IInterface#007877b1`
- `Substitute.IInterface:007877b1`
- `Substitute.IInterface|007877b1`
- ``Substitute.IInterface`007877b1``
- `Substitute.IInterface@007877b1`
- `Substitute.IInterface/007877b1`
- `Substitute.IInterface\007877b1`
- `Substitute.IInterface_007877b1`
- `Substitute.IInterface-007877b1`
- `Substitute.IInterface.007877b1`
- `Substitute.IInterface$007877b1`
- `Substitute.IInterface!007877b1`
- `Substitute.IInterface%007877b1`
- `Substitute.IInterface^007877b1`
- `Substitute.IInterface*007877b1`
- `Substitute.IInterface:#007877b1`
- `Substitute.IInterface[007877b1]`

Personally, I like the first 4 options, but would choose between `Substitute.IInterface:007877b1` and `Substitute.IInterface|007877b1`.

Here how the message looks with pipes:
```
  Expected: some item equal to <Substitute.IAdviseSink|03dc30f2>
  But was:  < <Substitute.IAdviseSink|03624f36>, <Substitute.IAdviseSink|03ac360b> >
```

Proceed with pipes for now, let me know if you see better option.

Notice, we cannot use `proxy.GetType().Name` to get ID, because it will always return `ObjectProxy` no matter what we mock. It's less useful than the primary type name.